services/http/api/lib/auth/auth_methods.ts

```diff
@@ -65,67 +66,137 @@ async function getRemoteId(token: BinaryLike) {
     }
 }
 
+async function matchClientCreatorRole(db: Session, key: ApiKeyInstance): Promise<boolean> {
+	debug("Finding client creator role");
+	const matches = await key.findRelationships({
+		alias: 'ClientCreatorRole',
+        session: db,
+		limit: 1
+	});
+	debug("Client creator matches: ", matches);
+	return matches.length >= 1;
+}
+
+async function matchDataExporterRole(db: Session, models: allModels, key: ApiKeyInstance, source: string)
+: Promise<boolean> {
+	const matches = await key.findRelationships({
+		alias: 'DataExporterRole',
+        session: db,
+	});
+	return await findOne(matches, async (m) => {
+		const sources = await m.target.findRelationships({
+			alias: 'Source',
+			where: {
+				target: {
+					name: source
+				},
+				relationship: {},
+			},
+			session: db,
+			limit: 1
+		});
+		return sources.length == 1;
+	}) != null;
+}
+
+async function matchClientReaderRole(db: Session, models: allModels,
+					user: UserIdentityInstance, clientId: string): Promise<boolean> {
+	const matches = await user.findRelationships({
+		alias: 'ClientReaderRole',
+		where: {
+			target: {},
+			relationship: {}
+		},
+		session: db
+	});
+	debug(`Client ID: ${clientId}`);
+	return await findOne(matches, async(m) => {
+		const clients = await m.target.findRelationships({
+			alias: 'User',
+			where: {
+				target: {
+					uid: clientId,
+				},
+				relationship: {}
+			},
+			session: db,
+			limit: 1,
+		});
+		return clients.length == 1;
+	}) != null;
+}
+
+
 export type AuthResult = DatabaseProcedure<boolean>;
 export type AuthMethod = (req: Request, auth: BinaryLike) => AuthResult;
 
+
 export const authMethods: AuthMethod[] = [
     // KEY AUTHENTICATION
     (req: Request, auth: BinaryLike) => async (db: Session, models: allModels) => {
-    const ident: IdentityInstance = await models.identity.findOne({
-        where: {
-            type: 'key',
-            check: basicHash(auth)
-        },
-        session: db
-    });
-    if (ident && ident.__existsInDatabase) {
-        const perms = await ident.findRelationships({
-            alias: "Resource",
-            where: {
-                target: {
-                    path: req.path,
-                },
-                relationship: {
-                    method: req.method
-                }
-            },
-            session: db,
-            limit: 1
-        });
-        return perms.length == 1;
-    } else {
-        return false;
-    }
+		debug("Key auth");
+		const key: ApiKeyInstance = await models.auth.apiKey.findOne({
+			where: {
+				hash: basicHash(auth),
+			},
+			session: db
+		});
+		if (!(key?.__existsInDatabase)) {
+			debug("Key DNE");
+			return false;
+		}
+		debug(`method: ${req.method}`);
+		switch (req.method) {
+			case 'POST':
+				switch (req.baseUrl) {
+					case '/client':
+						return await matchClientCreatorRole(db, key);
```

Comment:
I'm wondering if we can generalize these functions so we don't have to explicitly define them for a particular use case but rather have it so that, for example, a role can have a "scope" (e.g. "all", "self", "clients") associated with each object type (e.g. "clients" or "measurements") and that way we can just match the role, the object type, and the scope in a general function to figure out if the user has the permission requested or not. In such a case, we probably don't even need all these case statements as we could directly pass in the object type inferred by the URL into that general function.